### PR TITLE
jDataView 3.0.0

### LINKED ISSUE #89: Remove old fallbacks
When this library was created, DataView was a new WebGL extension available only in some browsers, but nowadays it's universally supported in JS engines.

We should remove old polyfills and let jDataView just take care of helper functionality (seeking, new types, etc.).

If someone wants to take this, please let me know.

### PR DESCRIPTION
**Pull request by zadeviggers**
See #89 for more motivation info

## Changes
- `jDataView` is now a ES6 class, not a constructor function
- To match with `DataView`, the `setInt64`, `getInt64`, `setUInt64`,  and `setUInt64` are now `setBigInt64`, `getBigInt64`, `setBigUInt64`, and `setBigUInt64`.
  - These also use `BigInt`s rather than regular numbers now, again for parity with `DataView`.
- NodeJS `Buffer`s aren't supported any more (can we even support these if it's driven by native `DataView`?)

## Tasks
- [x] Use ES6 modules & classes
- [x] Remove old pollyfills
- [x] Use Vite as the build pipeline
- [x] Use Vitest for testing
- [x] Split tests into separate files
- [x] Change all uses of `var` to `let` or `const`
- [x] Change usage of `arguments` to use `...args` instead
- [x] Re-add `writeXXX` methods
- [x] Stop extending DataView
- [x] Fix broken tests
- [x] Re-setup CI
- [x] Set up browser testing (is this needed?)
- [x] All tests passing
- [x] Write JSDoc docstrings for all the methods
- [x] Update the changelog
- [x] Update the readme
- [ ] Update github wiki
- [x] Test using in Node project
- [x] Test in Deno project
- [ ] Test using in a browser project
- [ ] Publish
- [ ] Close #89

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+* **September 26 2023**
+  * Version 3.0.0
+  * jDataView no longer polyfills DataView, it uses the native implementation. This should make it faster for some operations.
```
Comment (RReverser) [inline CHANGELOG.md:3]: This claim seems dubious - didn't jDataView already use DataView when available? If so, how does removing fallbacks make it faster? I'd rather rephrase this to say that the size of the bundle is smaller or something.

File: package.json [hunk 2]
```diff
@@ -12,52 +42,18 @@
     "write",
     "manipulation"
   ],
+  "license": "WTFPL",
```
Comment (RReverser) [inline package.json:45]: Please revert license change.

File: vite.config.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+export default defineConfig({
+	build: {
+		lib: {
+			// Could also be a dictionary or array of multiple entry points
+			entry: resolve(__dirname, "src/jdataview.js"),
+			name: "jDataView",
+			// the proper extensions will be added
+			fileName: "jdataview",
+			formats: ["es", "cjs", "umd"]
```
Comment (RReverser) [inline vite.config.js:14]: If we're doing major bump anyway, these days I'd rather keep only ESM output. Just needs careful testing that our package.json works correctly in both browser and Node.

Comment (zadeviggers) [reply vite.config.js]: Okay, I've got it set to just build es modules. I'm worried this might have some issues using in node, but I'll have to test that later.

File: package.json (hunk 2 above)
Comment (zadeviggers) [reply package.json:44]: I thought I left it the same as it was? I just moved the license field with the other metadata fields.

Yeah here's the license from before I forked it: https://github.com/jDataView/jDataView/blob/0592e6d6cd31b26cf1b3dcacd69b16f29c5f5bca/package.json#L75

Comment (zadeviggers) [reply package.json]: Oh I just found #86 - the branch I forked from must have been a bit outdated. I've changed it.

File: CHANGELOG.md (hunk 1 above)
Comment (zadeviggers) [reply CHANGELOG.md:3]: Yeah, I think I misunderstood some of how it used to work. Changed!

**Comment by zadeviggers** [comment]
@RReverser other than a few more docstrings and some integration testing, this should be good to go!

File: src/jdataview.js [hunk 4]
```diff
@@ -289,29 +363,41 @@ export class jDataView {
...
+	"Int16": 2, "Uint16": 2,
+	"Int8": 1, "Uint8": 1
+};
 // Encapsulate all the built-in methods
-for (const type of builtInTypes) {
+for (const type in builtInTypeBytes) {
+	const typeByteLength = builtInTypeBytes[type];
 	// Getters
 	jDataView.prototype["get" + type] = function (byteOffset, littleEndian) {
 		littleEndian = defined(littleEndian, this.littleEndian);
```
Comment (RReverser) [inline src/jdataview.js:378]: Note: in most (but probably not all) places those `defined` helpers can be replaced with ES6 param defaults now.

File: src/jdataview.js [hunk 5]
```diff
@@ -320,158 +96,115 @@ var proto = jDataView.prototype = {
...
 		return toArray ? arrayFrom(result) : result;
-	},
+	}
 
-	_setBytes: function (byteOffset, bytes, littleEndian) {
-		var length = bytes.length;
+	#setBytes(byteOffset, bytes, littleEndian) {
+		const length = bytes.length;
 
 		// needed for Opera
```
Comment (RReverser) [inline src/jdataview.js:147]: Likely not needed anymore.

File: src/jdataview.js [hunk 6]
```diff
@@ -480,354 +213,199 @@ var proto = jDataView.prototype = {
...
-			this.setUint32(byteOffset + parts[partName], value[partName], littleEndian);
-		}
+	setSigned(byteOffset, value, bitLength) {
+		return this.setUnsigned(byteOffset, value, bitLength);
+	}
 
-		this._offset = byteOffset + 8;
-	},
-
-	setInt64: function (byteOffset, value, littleEndian) {
```
Comment (RReverser) [inline src/jdataview.js:738]: Let's keep those methods. While BigInt is the way to go for large numbers nowadays, there are still lots of scenarios where it's convenient to allow writing numbers (up to 2^53) as if they were 64 bit.

Since new methods are called `setBigInt64`, there should be no name clash anyway.

File: src/jdataview.js (hunk 4 above)
Comment (RReverser) [reply src/jdataview.js:378]: The rest can be replaced with `??` or `??=` operators, although it depends which browsers we want to cover... I guess those who need it can always transpile anyway.

File: test/test-helpers.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    value = Array.prototype.slice.call(value);
+    assert.deepEqual(
+        value,
+        expected,
+        message || "[" + value + "] != [" + expected + "]"
+    );
+}
+
+function isNaN(obj) {
+    return Number.isNaN(obj) || (typeof obj === 'number' && obj.toString() === 'NaN');
```
Comment (RReverser) [inline test/test-helpers.js:29]: I'm guessing this helper is no longer needed - `Number.isNaN` should be enough.

File: src/jdataview.js (hunk 4 above)
Comment (zadeviggers) [reply src/jdataview.js:378]: Yeah they'll get automagically transpiled by Vite. I entirely just forgot to swap them out!

File: src/jdataview.js (hunk 6 above)
Comment (zadeviggers) [reply src/jdataview.js:738]: Alright I've added them. They just convert the input to a bigint or return value to a number and pass it off to the bigint methods.

When I was adding extra tests for the new methods, I removed the old compareInt64 helper function, which converted them back to regular numbers before comparing them. However, In doing so, one of the tests for `bigUint64` (https://github.com/zadeviggers/jDataView/blob/3df8b2b0f0bac9933dca152b1df7c417398b6c2f/test/getters.test.js#L210) started failing. Looking at it, it seems like the test was relying on loss of precision to pass. Should I just update it, or am I on the wrong track here?

**Comment by zadeviggers** [comment]
@RReverser do you know if there's a way to document methods added via dynamic prototype extension with jsdoc? I can't find anything online.

**Comment by zadeviggers** [comment]
I've changed the types to use a `.d.ts` file, so that I can use mapped types to generate the types for the dynamically added methods. This means they'll show up in autocomplete for everyone and typescript users will get type checking.

File: src/jdataview.d.ts [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+		byteOffset?: number,
+		byteLength?: number,
+		littleEndian?: boolean
+	);
+
+	// The `getBytes` method has an extra argument that changes the return type
+	// from a Uint8Array to a number[] if it's true
+	getBytes<
+		T extends boolean | undefined,
+		R extends T extends true ? number[] : Uint8Array
```
Comment (RReverser) [inline src/jdataview.d.ts:87]: Putting it here implies R is a generic and you can customise it; but you can't - it's a concrete type so `T extends ...` should be just put directly as the return type.

That said, I'm not sure there is still value in supporting pure array here in the upgrade - since we can expect Uint8Array to be available everywhere just like DataView, let's return it unconditionally. If someone wants to convert, they can still do so via `Array.from`.

File: src/jdataview.js [hunk 9]
```diff
@@ -1,833 +1,474 @@
...
+	}
 
-			this._offset = byteOffset + byteLength;
-			return this.buffer.toString(encoding || 'binary', this.byteOffset + byteOffset, this.byteOffset + this._offset);
-		}
-		var bytes = this._getBytes(byteLength, byteOffset, true);
+	/**
+	 * Read a string using the specified encoding, or binary if unspecified
+	 */
+	getString(byteLength, byteOffset, encoding) {
```
Comment (RReverser) [inline src/jdataview.js:226]: Let's simplify string ops to assume TextEncoder/TextDecoder are also always available.

Comment (zadeviggers) [reply src/jdataview.js:225]: Do we want to keep binary as a separate option? I'm not super familiar with TextEncoder/TextDecoder, but from glancing at the MDN page, it doesn't seem to support it.

Comment (RReverser) [reply src/jdataview.js]: Yeah binary would still need to be supported manually, I'm only talking about custom fallback for UTF-8.

File: src/jdataview.js [hunk 10]
```diff
@@ -1,833 +1,465 @@
...
 
-	// Helpers
+	// Setters and getters
 
-	_getBytes: function (length, byteOffset, littleEndian) {
-		littleEndian = defined(littleEndian, this._littleEndian);
-		byteOffset = defined(byteOffset, this._offset);
-		length = defined(length, this.byteLength - byteOffset);
+	#getBytes(length, byteOffset, littleEndian) {
+		littleEndian ??= this.littleEndian;
```
Comment (RReverser) [inline src/jdataview.js:172]: Nit: all these `??=` that can be inlined as param defaults should be inlined. Let's keep the new operator only for cases where it's really necessary.

File: src/jdataview.js [hunk 11]
```diff
@@ -1,833 +1,465 @@
...
+		if (buffer instanceof jDataView) {
+			const result = buffer.slice(byteOffset, byteOffset + byteLength);
+			result.littleEndian = littleEndian ?? result.littleEndian;
+			return result;
+		}
+
+		/**
+		 * The internal `ArrayBuffer` that jDataView is a view on
+		 */
+		this.buffer = wrapBuffer(buffer); // Convert strings, arrays, etc to `ArrayBuffer`s
```
Comment (RReverser) [inline src/jdataview.js:34]: Now that we always wrap DataView, it seems we duplicate those 3 properties between our instance and the nested DataView.

Assuming we never modify them (and I suppose not, since we have separate `#bytePointer`), let's get them from the native instance when necessary instead to reduce duplication.

Since we still want to conform to DataView interface, it probably means declaring getters for these 3 properties that just return inner values instead.

**Comment by RReverser** [comment]
> * NodeJS `Buffer`s aren't supported any more (can we even support these if it's driven by native `DataView`?)

That sounds odd btw. Node.js `Buffer` is a subclass of `Uint8Array` nowadays, so as long as we support `Uint8Array`, it should work too. Why does it break?

File: src/jdataview.d.ts [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+	| Int16Array
+	| Uint16Array
+	| Int32Array
+	| Uint32Array
+	| Float32Array
+	| Float64Array
+	| BigInt64Array
+	| BigUint64Array;
+
+type Bufferish = string | number | ArrayBuffer | TypedArray | ArrayLike<number>;
```
Comment (RReverser) [inline src/jdataview.d.ts:67]:
```suggestion
type Bufferish = string | number | ArrayBuffer | ArrayLike<number>;
```

I believe TypedArray is covered by ArrayLike<number>.

File: src/jdataview.d.ts [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+
+declare class jDataView {
```
Comment (RReverser) [inline src/jdataview.d.ts:69]:
```suggestion
declare class jDataView implements DataView {
```

might be helpful as a self-check that we are still a drop-in replacement. E.g. if that interface above declared some incompatible signature, TS would complain here.

Right now it seems to complain only about missing `Symbol.toStringTag` - we should probably provide a custom implementation there.

File: src/jdataview.js [hunk 14]
```diff
@@ -1,833 +1,471 @@
...
-	_setBytes: function (byteOffset, bytes, littleEndian) {
-		var length = bytes.length;
-
-		// needed for Opera
-		if (length === 0) {
-			return;
-		}
+		return littleEndian || length <= 1
+			? result
+			: Array.from(result).reverse();
```
Comment (RReverser) [inline src/jdataview.js:186]: Won't this return an inconsistent type now?

It should use `.reverse()` directly on `Uint8Array` to prevent either `Uint8Array` or `Array` being returned depending on parameter.

But now I'm also concerned that tests don't catch this - the comparisons should be verifying type of the input, and I believe deepEqual does by default, so the problem is that there is likely some extra casting before comparing values.

File: src/jdataview.js [hunk 15]
```diff
@@ -1,833 +1,471 @@
...
+		littleEndian = this.littleEndian
+	) {
+		const length = bytes.length;
 
-		this._checkBounds(byteOffset, length);
+		this.#checkBounds(byteOffset, length);
 
 		if (!littleEndian && length > 1) {
-			bytes = arrayFrom(bytes, true).reverse();
+			bytes = Array.from(bytes).reverse();
```
Comment (RReverser) [inline src/jdataview.js:199]: Same here.

File: src/jdataview.js (hunk 14 above)
Comment (RReverser) [reply src/jdataview.js:186]: It would be nice to integrate at least minimal TS checking (via `checkJS`) so that it would catch basic type issues. I know it won't work for the dynamic extension to prototype, but even if we left some `any` here and there it's still better than nothing.

File: test/getters.test.js [hunk 16]
```diff
@@ -0,0 +1,300 @@
...
+				const offset = contextView.tell();
+				const realValue = contextView["get" + type].apply(
+					contextView,
+					args
+				);
+
+				check(
+					realValue,
+					value,
+					"get" +
```
Comment (RReverser) [inline test/getters.test.js:35]: Wow this pre-ES6 code is something when formatted. Nit, but since you converted the rest to ES6, can this be turned into template literal for readability please?

File: test/getters.test.js [hunk 17]
```diff
@@ -0,0 +1,300 @@
...
+									.join(", ") +
+							  "]"
+							: "")
+				);
+			});
+		});
+	}
+
+	testGetters("Bytes", [
+		{ args: [2, 1], value: [0xfe, 0xfd], check: compareBytes },
```
Comment (RReverser) [inline test/getters.test.js:58]: I guess if we just removed the `compareBytes` helper and turned `value` here into `Uint8Array.of(0xfe, 0xfd)`, it would catch those type mismatches?

Comment (RReverser) [reply test/getters.test.js:58]: Or idk change `compareBytes` to auto-convert argument via `Uint8Array.from`. Either one is fine.

File: test/getters.test.js [hunk 18]
```diff
@@ -0,0 +1,300 @@
...
+
+	test(
+		"Big String",
+		function () {
+			const view = new jDataView(2000000);
+			assert.doesNotThrow(function () {
+				view.getString();
+			});
+		},
+		{ timeout: 5000 }
```
Comment (RReverser) [inline test/getters.test.js:101]: Why does this need timeout? It would always throw synchronously if at all.

File: src/helpers.js [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+		case "number":
+			buffer = new Uint8Array(buffer).buffer;
+
+			break;
+
+		case "string":
+			buffer = getCharCodes(buffer);
+		/* falls through */
+		default:
+			if (!(buffer instanceof ArrayBuffer)) {
```
Comment (RReverser) [inline src/helpers.js:22]: I think we should accept SharedArrayBuffer here too nowadays. The only problem is that it's conditionally defined depending on COOP/COEP headers, so it needs extra `typeof SharedArrayBuffer !== 'undefined`` check before `instanceof` so that it doesn't throw.

File: src/helpers.js [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+export function wrapBuffer(buffer) {
+	switch (typeof buffer) {
+		case "number":
+			buffer = new Uint8Array(buffer).buffer;
+
+			break;
+
+		case "string":
+			buffer = getCharCodes(buffer);
+		/* falls through */
```
Comment (RReverser) [inline src/helpers.js:20]: If `getCharCodes` now creates its own `Uint8Array`, there's no point in falling through to those extra checks. Let's just do `buffer = getCharCodes(buffer).buffer;` here and `break`.

File: src/helpers.js [hunk 21]
```diff
@@ -0,0 +1,27 @@
+export function getCharCodes(string) {
+	const codes = new Uint8Array(string.length);
+
+	for (let i = 0, length = string.length; i < length; i++) {
+		codes[i] = string.charCodeAt(i) & 0xff;
```
Comment (RReverser) [inline src/helpers.js:5]: Since destination is a `Uint8Array`, `& 0xff` is no longer necessary here - the typed array will take care of the wraparound.

Might be worth checking if there are more unnecessary bitcasts like this in the codebase.

File: src/jdataview.js [hunk 22]
```diff
@@ -1,833 +1,471 @@
...
+	 * Read a string using the specified encoding, or binary if unspecified
+	 */
+	getString(byteLength, byteOffset, encoding = "binary") {
+		const bytes = this.#getBytes(byteLength, byteOffset, true);
 		// backward-compatibility
-		encoding = encoding === 'utf8' ? 'utf-8' : (encoding || 'binary');
-		if (TextDecoder && encoding !== 'binary') {
-			return new TextDecoder(encoding).decode(this._isArrayBuffer ? bytes : new Uint8Array(bytes));
+		if (encoding !== "binary") {
+			return new TextDecoder(encoding).decode(bytes);
```
Comment (RReverser) [inline src/jdataview.js:230]: Note: creating instances of TextEncoder / TextDecoder can be relatively slow. Create those once in some variable outside the class and reuse instead.

File: src/jdataview.js [hunk 23]
```diff
@@ -1,833 +1,471 @@
...
+		wideValue &= ~(~(-1 << bitLength) << -this.#bitOffset); // clearing bit range before binary "or"
+		wideValue |=
+			(bitLength < 32 ? value & ~(-1 << bitLength) : value) <<
+			-this.#bitOffset; // setting bits
 
-	_set64: function (Type, byteOffset, value, littleEndian) {
-		if (typeof value !== 'object') {
-			value = Type.fromNumber(value);
+		for (let i = data.bytes.length - 1; i >= 0; i--) {
+			data.bytes[i] = wideValue & 0xff;
```
Comment (RReverser) [inline src/jdataview.js:326]: Here's another place where `& 0xff` is no longer necessary.

File: src/jdataview.js [hunk 24]
```diff
@@ -1,833 +1,471 @@
...
-		return isReadAction ? this.buffer['read' + nodeName](byteOffset) : this.buffer['write' + nodeName](value, byteOffset);
-	};
-}
+	// Setters
+	jDataView.prototype["set" + type] = function (
+		byteOffset = this.tell(),
+		value,
+		littleEndian = this.littleEndian
+	) {
+		if (typeof this === "undefined") {
```
Comment (RReverser) [inline src/jdataview.js:434]: Hm I don't think this check is worth it here and below. Most JS devs know that you can't just take `obj.method` and call it later without the instance. And if not, the strict mode will catch it pretty quickly for them as `this.seek` below will fail with `this === undefined`.

Comment (zadeviggers) [reply src/jdataview.js]: Okay, just something I ran into when testing in a node / deno project.

File: src/jdataview.js (hunk 22 above)
Comment (zadeviggers) [reply src/jdataview.js:230]: Wouldn't we need to create one for each encoding then?

File: src/jdataview.js (hunk 14 above)
Comment (zadeviggers) [reply src/jdataview.js]: Wow, I found a bunch of tiny little errors thanks to checkJS!

File: test/getters.test.js (hunk 17 above)
Comment (zadeviggers) [reply test/getters.test.js:50]: `compareBytes` is used in quite a few places, so I went with the latter.

File: test/getters.test.js (hunk 18 above)
Comment (zadeviggers) [reply test/getters.test.js:93]: I have a vague recollection form when I was converting the tests that it had a timeout for some reason.

My best guess is there might have been some old engine bug that caused the thread to hang on that call sometimes?

Anyway, seems like it's not useful now.

File: src/jdataview.js (hunk 22 above)
Comment (RReverser) [reply src/jdataview.js:233]: Yeah but we only used to support UTF-8, I suppose we can keep it that way. TextEncoder doesn't support anything else either.

**Review by RReverser — APPROVED** [review]
Ok I'm going to merge as it is, this got big enough as it is and if there are more issues, we can update in other PRs. Thanks a lot, this is great work!

**Comment by zadeviggers** [comment]
Awesome, glad to see it!

**Comment by zadeviggers** [comment]
Looks like you still need to do an npm publish. The readme could also use a bit of a makeover, but that's not urgent at all.

**Comment by RReverser** [comment]
Yeah it's only merged for now.

As you noted, this probably also needs a wiki update and would be good to test this with some real demos, in particular with https://jdataview.github.io/jBinary.Repo/demo/ which uses old library from gh-pages branch of this repo.

**Comment by RReverser** [comment]
Ok I'm sorry I thought I'll get around to update jBinary soon-ish but clearly not having enough time.

If you could update the docs (the wiki), I'm happy to cut a release. I sent you an invite with writing rights which should allow you access to editing the wiki.

**Comment by zadeviggers** [comment]
Yep, I'm happy to update the wiki. I'm about to leave for a week, but I'll see if I can get everything done this weekend.
I'll also do a once-over of the readme in another PR.
